File: go/mechanisms/evm/batched/server/channel_manager.go
package server

import (
	"context"
	"encoding/json"
	"fmt"
	"math/big"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	x402 "github.com/x402-foundation/x402/go"
	"github.com/x402-foundation/x402/go/mechanisms/evm"
	"github.com/x402-foundation/x402/go/mechanisms/evm/batched"
)

// ChannelManagerConfig holds the dependencies for a channel manager.
type ChannelManagerConfig struct {
	Scheme      *BatchedEvmScheme
	Facilitator x402.FacilitatorClient
	Network     x402.Network
}

// AutoSettlementConfig configures auto claim/settle/refund behavior.
type AutoSettlementConfig struct {
	// ClaimIntervalSecs claims on a fixed interval.
	ClaimIntervalSecs int
	// ClaimOnIdleSecs claims when channels are idle for this long.
	ClaimOnIdleSecs int
	// ClaimThreshold claims when total claimable exceeds this amount.
	ClaimThreshold string
	// MaxClaimsPerBatch limits the number of vouchers per claim tx.
	MaxClaimsPerBatch int
	// ClaimOnWithdrawal claims all when a withdrawal is detected.
	ClaimOnWithdrawal bool
	// SettleIntervalSecs settles on a fixed interval.
	SettleIntervalSecs int
	// SettleThreshold settles when total claimed exceeds this amount.
	SettleThreshold string
	// RefundOnIdleSecs cooperatively refunds channels idle for this long.
	RefundOnIdleSecs int
	// TickSecs is how often the manager evaluates triggers. Defaults to 5.
	TickSecs int
	// OnClaim is called after a successful claim.
	OnClaim func(ClaimResult)
	// OnSettle is called after a successful settle.
	OnSettle func(SettleResult)
	// OnRefund is called after a successful refund.
	OnRefund func(RefundResult)
	// OnError is called when any auto-settlement operation fails.
	OnError func(error)
}

// ClaimResult holds the result of a claim operation.
type ClaimResult struct {
	Vouchers    int
	Transaction string
}

// SettleResult holds the result of a settle operation.
type SettleResult struct {
	Transaction string
}

// RefundResult holds the result of a refund operation.
type RefundResult struct {
	Channels    []string
	Transaction string
}

// BatchedChannelManager handles auto-settlement of batched payment channels.
type BatchedChannelManager struct {
	scheme      *BatchedEvmScheme
	facilitator x402.FacilitatorClient
	network     x402.Network

	// Auto-settlement state
	ticker         *time.Ticker
	stopCh         chan struct{}
	lastClaimTime  time.Time
	lastSettleTime time.Time
	pendingSettle  bool
	config         AutoSettlementConfig
	running        bool
	tickInProgress int32 // atomic

	mu sync.Mutex
}

// NewBatchedChannelManager creates a new channel manager.
func NewBatchedChannelManager(config ChannelManagerConfig) *BatchedChannelManager {
	return &BatchedChannelManager{
		scheme:      config.Scheme,
		facilitator: config.Facilitator,
		network:     config.Network,
	}
}

// GetClaimableVouchersOpts filters claimable vouchers by idle time.
type GetClaimableVouchersOpts struct {
	IdleSecs int // Filter sessions idle for at least this many seconds
}

// GetClaimableVouchers returns voucher claims ready for on-chain settlement.
//
// A voucher is claimable when its `chargedCumulativeAmount` exceeds what has
// already been claimed on-chain. An optional `IdleSecs` filter skips sessions
// that received a request within the last `IdleSecs` seconds.
func (m *BatchedChannelManager) GetClaimableVouchers(opts *GetClaimableVouchersOpts) ([]batched.BatchedVoucherClaim, error) {
	sessions, err := m.scheme.storage.List()
	if err != nil {
		return nil, err
	}

	now := time.Now().UnixMilli()
	claims := make([]batched.BatchedVoucherClaim, 0)

	for _, session := range sessions {
		if opts != nil && opts.IdleSecs > 0 {
			idleMs := now - session.LastRequestTimestamp
			if idleMs < int64(opts.IdleSecs)*1000 {
				continue
			}
		}

		charged, _ := new(big.Int).SetString(session.ChargedCumulativeAmount, 10)
		claimed, _ := new(big.Int).SetString(session.TotalClaimed, 10)
		if charged == nil || claimed == nil {
			continue
		}
		if charged.Cmp(claimed) <= 0 {
			continue
		}

		claims = append(claims, batched.BatchedVoucherClaim{
			Voucher: struct {
				Channel            batched.ChannelConfig `json:"channel"`
				MaxClaimableAmount string                `json:"maxClaimableAmount"`
			}{
				Channel:            session.ChannelConfig,
				MaxClaimableAmount: session.SignedMaxClaimable,
			},
			Signature:    session.Signature,
			TotalClaimed: session.ChargedCumulativeAmount,
		})
	}

	return claims, nil
}

// GetWithdrawalPendingSessions returns sessions that have a pending payer-initiated
// withdrawal (withdrawRequestedAt > 0).
func (m *BatchedChannelManager) GetWithdrawalPendingSessions() ([]*ChannelSession, error) {
	sessions, err := m.scheme.storage.List()
	if err != nil {
		return nil, err
	}
	var result []*ChannelSession
	for _, session := range sessions {
		if session.WithdrawRequestedAt > 0 {
			result = append(result, session)
		}
	}
	return result, nil
}

// Claim collects and claims outstanding vouchers.
type ClaimOptions struct {
	MaxClaimsPerBatch int
	IdleSecs          int
}

func (m *BatchedChannelManager) Claim(ctx context.Context, opts *ClaimOptions) ([]ClaimResult, error) {
	idleSecs := 0
	maxClaims := 50
	if opts != nil {
		if opts.IdleSecs > 0 {
			idleSecs = opts.IdleSecs
		}
		if opts.MaxClaimsPerBatch > 0 {
			maxClaims = opts.MaxClaimsPerBatch
		}
	}

	claims, err := m.GetClaimableVouchers(&GetClaimableVouchersOpts{IdleSecs: idleSecs})
	if err != nil {
		return nil, fmt.Errorf("failed to get claimable vouchers: %w", err)
	}

	if len(claims) == 0 {
		return nil, nil
	}

	// Batch claims
	var results []ClaimResult
	for i := 0; i < len(claims); i += maxClaims {
		end := i + maxClaims
		if end > len(claims) {
			end = len(claims)
		}
		batch := claims[i:end]

		result, err := m.executeClaim(ctx, batch)
		if err != nil {
			return results, fmt.Errorf("claim batch failed: %w", err)
		}
		results = append(results, *result)
	}

	return results, nil
}

// Settle transfers claimed funds to the receiver.
func (m *BatchedChannelManager) Settle(ctx context.Context) (*SettleResult, error) {
	settle := &batched.BatchedSettlePayload{
		Type:     "settle",
		Receiver: m.scheme.receiverAddress,
		Token:    m.getToken(),
	}
	resp, err := m.facilitatorSettle(ctx, settle.ToMap())
	if err != nil {
		return nil, err
	}
	return &SettleResult{Transaction: resp.Transaction}, nil
}

// facilitatorSettle marshals the (payload, requirements) pair this manager uses
// for its claim/settle/refund calls and forwards them to the facilitator.
func (m *BatchedChannelManager) facilitatorSettle(ctx context.Context, payloadMap map[string]interface{}) (*x402.SettleResponse, error) {
	payloadBytes, err := json.Marshal(map[string]interface{}{
		"x402Version": 2,
		"payload":     payloadMap,
		"accepted": map[string]interface{}{
			"scheme":  batched.SchemeBatched,
			"network": string(m.network),
		},
	})
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	requirementsBytes, err := json.Marshal(map[string]interface{}{
		"scheme":  batched.SchemeBatched,
		"network": string(m.network),
		"payTo":   m.scheme.receiverAddress,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to marshal requirements: %w", err)
	}

	return m.facilitator.Settle(ctx, payloadBytes, requirementsBytes)
}

// Refund cooperatively refunds the given channels. For each channel, claims
// any outstanding vouchers and refunds the unclaimed balance back to the payer.
func (m *BatchedChannelManager) Refund(ctx context.Context, channelIds []string) (*RefundResult, error) {
	if len(channelIds) == 0 {
		return nil, nil
	}

	var refundedChannels []string
	var lastTxHash string

	for _, channelId := range channelIds {
		normalizedId := batched.NormalizeChannelId(channelId)
		session, err := m.scheme.storage.Get(normalizedId)
		if err != nil || session == nil {
			continue
		}

		// Build claim entry from current session state
		claimEntry := batched.BatchedVoucherClaim{
			Voucher: struct {
				Channel            batched.ChannelConfig `json:"channel"`
				MaxClaimableAmount string                `json:"maxClaimableAmount"`
			}{
				Channel:            session.ChannelConfig,
				MaxClaimableAmount: session.SignedMaxClaimable,
			},
			Signature:    session.Signature,
			TotalClaimed: session.ChargedCumulativeAmount,
		}

		// Calculate refund amount: balance - chargedCumulativeAmount
		balance, _ := new(big.Int).SetString(session.Balance, 10)
		charged, _ := new(big.Int).SetString(session.ChargedCumulativeAmount, 10)
		if balance == nil || charged == nil {
			continue
		}
		refundAmount := new(big.Int).Sub(balance, charged)
		if refundAmount.Sign() <= 0 {
			continue
		}

		// Pre-sign if server has authorizerSigner; otherwise facilitator auto-signs.
		nonce := fmt.Sprintf("%d", session.RefundNonce)
		refund := &batched.BatchedEnrichedRefundPayload{
			Type:          "refund",
			ChannelConfig: session.ChannelConfig,
			Voucher: batched.BatchedVoucherFields{
				ChannelId:          normalizedId,
				MaxClaimableAmount: session.SignedMaxClaimable,
				Signature:          session.Signature,
			},
			Amount:      refundAmount.String(),
			RefundNonce: nonce,
			Claims:      []batched.BatchedVoucherClaim{claimEntry},
		}

		if m.scheme.receiverAuthorizerSigner != nil {
			authSig, err := m.scheme.SignRefund(ctx, normalizedId, refundAmount.String(), nonce, string(m.network))
			if err != nil {
				return nil, fmt.Errorf("sign refund for %s: %w", normalizedId, err)
			}
			claimAuthSig, err := m.scheme.SignClaimBatch(ctx, []batched.BatchedVoucherClaim{claimEntry}, string(m.network))
			if err != nil {
				return nil, fmt.Errorf("sign claim batch for %s: %w", normalizedId, err)
			}
			refund.RefundAuthorizerSignature = evm.BytesToHex(authSig)
			refund.ClaimAuthorizerSignature = evm.BytesToHex(claimAuthSig)
		}

		resp, err := m.facilitatorSettle(ctx, refund.ToMap())
		if err != nil {
			return nil, fmt.Errorf("refund settle for %s: %w", normalizedId, err)
		}
		if !resp.Success {
			return nil, fmt.Errorf("refund settle for %s failed: %s — %s", normalizedId, resp.ErrorReason, resp.ErrorMessage)
		}

		lastTxHash = resp.Transaction
		refundedChannels = append(refundedChannels, normalizedId)

		// Delete the session after successful refund
		_ = m.scheme.storage.Delete(normalizedId)
	}

	if len(refundedChannels) == 0 {
		return nil, nil
	}

	return &RefundResult{
		Channels:    refundedChannels,
		Transaction: lastTxHash,
	}, nil
}

// ClaimAndSettle claims then settles.
func (m *BatchedChannelManager) ClaimAndSettle(ctx context.Context, opts *ClaimOptions) (*SettleResult, error) {
	_, err := m.Claim(ctx, opts)
	if err != nil {
		return nil, err
	}
	return m.Settle(ctx)
}

// Start begins auto-settlement with the given configuration.
func (m *BatchedChannelManager) Start(config AutoSettlementConfig) {
	m.mu.Lock()
	defer m.mu.Unlock()

	if m.running {
		return
	}

	m.config = config
	m.running = true
	m.lastClaimTime = time.Now()
	m.lastSettleTime = time.Now()

	tickSecs := config.TickSecs
	if tickSecs <= 0 {
		tickSecs = 5
	}

	m.ticker = time.NewTicker(time.Duration(tickSecs) * time.Second)
	m.stopCh = make(chan struct{})

	go func() {
		for {
			select {
			case <-m.ticker.C:
				m.tick()
			case <-m.stopCh:
				return
			}
		}
	}()
}

// Stop halts auto-settlement. If flush is true, performs a final ClaimAndSettle.
func (m *BatchedChannelManager) Stop(ctx context.Context, flush bool) error {
	m.mu.Lock()
	if !m.running {
		m.mu.Unlock()
		return nil
	}

	m.running = false
	m.ticker.Stop()
	close(m.stopCh)
	m.mu.Unlock()

	if flush {
		_, err := m.ClaimAndSettle(ctx, nil)
		return err
	}

	return nil
}

func (m *BatchedChannelManager) tick() {
	if !atomic.CompareAndSwapInt32(&m.tickInProgress, 0, 1) {
		return // Skip if previous tick is still running
	}
	defer atomic.StoreInt32(&m.tickInProgress, 0)

	ctx := context.Background()
	config := m.config
	now := time.Now()

	// Check claim triggers
	shouldClaim := false
	claimOpts := &ClaimOptions{}

	if config.MaxClaimsPerBatch > 0 {
		claimOpts.MaxClaimsPerBatch = config.MaxClaimsPerBatch
	}

	// Time-based claim
	if config.ClaimIntervalSecs > 0 && now.Sub(m.lastClaimTime) >= time.Duration(config.ClaimIntervalSecs)*time.Second {
		shouldClaim = true
	}

	// Idle-based claim
	if config.ClaimOnIdleSecs > 0 {
		claimOpts.IdleSecs = config.ClaimOnIdleSecs
		shouldClaim = true
	}

	// Withdrawal-based claim: if a payer initiated a withdrawal, claim their vouchers
	if config.ClaimOnWithdrawal {
		withdrawals, err := m.GetWithdrawalPendingSessions()
		if err == nil && len(withdrawals) > 0 {
			claimable, err := m.GetClaimableVouchers(nil)
			if err == nil {
				withdrawalPayers := make(map[string]bool)
				for _, w := range withdrawals {
					withdrawalPayers[strings.ToLower(w.ChannelConfig.Payer)] = true
				}
				for _, c := range claimable {
					if withdrawalPayers[strings.ToLower(c.Voucher.Channel.Payer)] {
						shouldClaim = true
						break
					}
				}
			}
		}
	}

	// Threshold-based claim
	if config.ClaimThreshold != "" {
		threshold, ok := new(big.Int).SetString(config.ClaimThreshold, 10)
		if ok {
			claims, err := m.GetClaimableVouchers(nil)
			if err == nil {
				total := big.NewInt(0)
				for _, claim := range claims {
					tc, _ := new(big.Int).SetString(claim.TotalClaimed, 10)
					if tc != nil {
						total.Add(total, tc)
					}
				}
				if total.Cmp(threshold) > 0 {
					shouldClaim = true
				}
			}
		}
	}

	if shouldClaim {
		results, err := m.Claim(ctx, claimOpts)
		if err != nil {
			if config.OnError != nil {
				config.OnError(fmt.Errorf("auto-claim failed: %w", err))
			}
		} else if len(results) > 0 {
			m.lastClaimTime = now
			m.pendingSettle = true
			if config.OnClaim != nil {
				for _, result := range results {
					config.OnClaim(result)
				}
			}
		}
	}

	// Check settle triggers
	shouldSettle := config.SettleIntervalSecs > 0 &&
		now.Sub(m.lastSettleTime) >= time.Duration(config.SettleIntervalSecs)*time.Second &&
		m.pendingSettle

	if shouldSettle {
		result, err := m.Settle(ctx)
		if err != nil {
			if config.OnError != nil {
				config.OnError(fmt.Errorf("auto-settle failed: %w", err))
			}
		} else if result != nil {
			m.lastSettleTime = now
			m.pendingSettle = false
			if config.OnSettle != nil {
				config.OnSettle(*result)
			}
		}
	}

	// Check refund triggers (idle channels)
	if config.RefundOnIdleSecs > 0 {
		sessions, err := m.scheme.storage.List()
		if err == nil {
			nowMs := now.UnixMilli()
			var refundChannelIds []string
			for _, session := range sessions {
				idleMs := nowMs - session.LastRequestTimestamp
				if idleMs >= int64(config.RefundOnIdleSecs)*1000 {
					refundChannelIds = append(refundChannelIds, session.ChannelId)
				}
			}
			if len(refundChannelIds) > 0 {
				result, err := m.Refund(ctx, refundChannelIds)
				if err != nil {
					if config.OnError != nil {
						config.OnError(fmt.Errorf("auto-refund failed: %w", err))
					}
				} else if result != nil && config.OnRefund != nil {
					config.OnRefund(*result)
				}
			}
		}
	}
}

func (m *BatchedChannelManager) executeClaim(ctx context.Context, claims []batched.BatchedVoucherClaim) (*ClaimResult, error) {
	// Pre-sign with the receiver's authorizer when configured; otherwise the
	// facilitator auto-signs via its own AuthorizerSigner.
	claim := &batched.BatchedClaimPayload{Type: "claim", Claims: claims}
	if m.scheme.receiverAuthorizerSigner != nil {
		sig, err := m.scheme.SignClaimBatch(ctx, claims, string(m.network))
		if err != nil {
			return nil, fmt.Errorf("failed to sign claim batch: %w", err)
		}
		claim.ClaimAuthorizerSignature = evm.BytesToHex(sig)
	}

	resp, err := m.facilitatorSettle(ctx, claim.ToMap())
	if err != nil {
		return nil, err
	}
	if !resp.Success {
		return nil, fmt.Errorf("claim settle failed: %s — %s", resp.ErrorReason, resp.ErrorMessage)
	}

	return &ClaimResult{
		Vouchers:    len(claims),
		Transaction: resp.Transaction,
	}, nil
}

func (m *BatchedChannelManager) getToken() string {
	sessions, err := m.scheme.storage.List()
	if err != nil || len(sessions) == 0 {
		return ""
	}
	return sessions[0].ChannelConfig.Token
}

File: python/x402/mechanisms/evm/batch_settlement/server/channel_manager.py
"""Async `BatchSettlementChannelManager` for periodic claim/settle/refund."""

from __future__ import annotations

import asyncio
from collections.abc import Awaitable
from typing import Any

from ..authorizer_signer import sign_claim_batch, sign_refund
from ..types import VoucherClaim
from ..utils import compute_channel_id
from .channel_manager_common import (
    AUTO_JOB_PRIORITY,
    AutoJob,
    AutoSettlementConfig,
    AutoSettlementContext,
    ChannelManagerConfig,
    ClaimChannelSelector,
    ClaimOptions,
    ClaimResult,
    RefundChannelSelector,
    RefundResult,
    SettleResult,
    build_claim_payment_payload,
    build_payment_requirements,
    build_refund_claims,
    build_refund_payment_payload,
    build_settle_payment_payload,
    format_facilitator_failure,
    get_claimable_vouchers_from_channels,
    get_idle_channels_for_refund,
    has_live_pending_request,
    now_ms,
)
from .storage import Channel


class BatchSettlementChannelManager:
    """Async channel manager for claim / settle / refund lifecycle."""

    def __init__(self, config: ChannelManagerConfig) -> None:
        self._scheme = config.scheme
        self._facilitator = config.facilitator
        self._receiver = config.receiver
        self._token = config.token
        self._network = config.network

        self._timer_tasks: dict[AutoJob, asyncio.Task[None]] = {}
        self._drain_task: asyncio.Task[None] | None = None
        self._last_claim_time = 0
        self._last_settle_time = 0
        self._pending_settle = False
        self._running = False
        self._pending_jobs: set[AutoJob] = set()
        self._draining_jobs = False
        self._auto_settle_config = AutoSettlementConfig()

    async def claim(self, opts: ClaimOptions | None = None) -> list[ClaimResult]:
        targets = await self._select_claim_targets(opts)
        max_per_batch = (opts.max_claims_per_batch if opts else None) or 100
        idle_secs = opts.idle_secs if opts else None
        return await self._claim_from_channels(
            targets, max_claims_per_batch=max_per_batch, idle_secs=idle_secs
        )

    async def settle(self) -> SettleResult:
        requirements = build_payment_requirements(self._receiver, self._token, self._network)
        payload = build_settle_payment_payload(requirements, self._receiver, self._token)
        response = await self._facilitator.settle(payload, requirements)
        if not response.success:
            raise RuntimeError(format_facilitator_failure("Settle", response))
        self._pending_settle = False
        return SettleResult(transaction=response.transaction)

    async def claim_and_settle(self, opts: ClaimOptions | None = None) -> dict[str, Any]:
        claims = await self.claim(opts)
        settle: SettleResult | None = None
        if claims:
            settle = await self.settle()
        return {"claims": claims, "settle": settle}

    async def refund(self, channel_ids: list[str] | None = None) -> list[RefundResult]:
        storage = self._scheme.get_storage()
        channels = storage.list()

        now = now_ms()
        if channel_ids is not None:
            requested = {c.lower() for c in channel_ids}
            channels = [c for c in channels if c.channel_id.lower() in requested]
        targets = [c for c in channels if not has_live_pending_request(c, now)]

        if not targets:
            return []
        return await self._refund_channels(targets)

    async def refund_idle_channels(self, idle_secs: int) -> list[RefundResult]:
        targets = get_idle_channels_for_refund(self._scheme.get_storage().list(), idle_secs)
        return await self._refund_channels(targets)

    async def get_claimable_vouchers(
        self, opts: dict[str, Any] | None = None
    ) -> list[VoucherClaim]:
        idle_secs = opts.get("idleSecs") if opts else None
        channels = self._scheme.get_storage().list()
        return get_claimable_vouchers_from_channels(channels, idle_secs)

    async def get_withdrawal_pending_sessions(self) -> list[Channel]:
        return [c for c in self._scheme.get_storage().list() if c.withdraw_requested_at > 0]

    def start(self, config: AutoSettlementConfig | None = None) -> None:
        if self._running:
            return
        now = now_ms()
        self._last_claim_time = now
        self._last_settle_time = now
        self._running = True
        self._auto_settle_config = config or AutoSettlementConfig()

        self._start_auto_timer("claim", self._auto_settle_config.claim_interval_secs)
        self._start_auto_timer("settle", self._auto_settle_config.settle_interval_secs)
        self._start_auto_timer("refund", self._auto_settle_config.refund_interval_secs)

    async def stop(self, *, flush: bool = False) -> None:
        self._running = False
        for task in list(self._timer_tasks.values()):
            task.cancel()
        self._timer_tasks.clear()
        self._pending_jobs.clear()

        drain = self._drain_task
        if drain is not None and not drain.done():
            drain.cancel()
            try:
                await drain
            except asyncio.CancelledError:
                pass
        self._drain_task = None

        if flush:
            opts = ClaimOptions(
                max_claims_per_batch=self._auto_settle_config.max_claims_per_batch,
                select_claim_channels=self._auto_settle_config.select_claim_channels,
            )
            await self.claim_and_settle(opts)

    def _start_auto_timer(self, job: AutoJob, interval_secs: int | None) -> None:
        if interval_secs is None:
            return

        async def loop() -> None:
            while self._running:
                await asyncio.sleep(interval_secs)
                if not self._running:
                    return
                self._enqueue_job(job)

        try:
            asyncio.get_running_loop()
        except RuntimeError as err:
            raise RuntimeError(
                "BatchSettlementChannelManager.start() requires a running asyncio "
                "event loop (e.g. call it from FastAPI lifespan startup, not at "
                "module import time)."
            ) from err
        self._timer_tasks[job] = asyncio.create_task(loop(), name=f"bs-{job}")

    def _enqueue_job(self, job: AutoJob) -> None:
        if not self._running:
            return
        self._pending_jobs.add(job)
        if not self._draining_jobs:
            self._drain_task = asyncio.create_task(self._drain_jobs())

    async def _drain_jobs(self) -> None:
        if self._draining_jobs:
            return
        self._draining_jobs = True
        try:
            while self._running and self._pending_jobs:
                next_job: AutoJob | None = next(
                    (j for j in AUTO_JOB_PRIORITY if j in self._pending_jobs),
                    None,
                )
                if next_job is None:
                    return
                self._pending_jobs.discard(next_job)
                await self._run_auto_job(next_job)
        finally:
            self._draining_jobs = False

    async def _run_auto_job(self, job: AutoJob) -> None:
        try:
            if job == "claim":
                await self._run_claim_job()
            elif job == "settle":
                await self._run_settle_job()
            elif job == "refund":
                await self._run_refund_job()
        except BaseException as err:  # noqa: BLE001
            on_error = self._auto_settle_config.on_error
            if on_error is not None:
                on_error(err)

    async def _run_claim_job(self) -> None:
        cfg = self._auto_settle_config
        targets = await self._select_claim_targets(
            ClaimOptions(select_claim_channels=cfg.select_claim_channels)
        )
        results = await self._claim_from_channels(
            targets,
            max_claims_per_batch=cfg.max_claims_per_batch or 100,
            idle_secs=None,
        )
        self._last_claim_time = now_ms()
        if cfg.on_claim is not None:
            for r in results:
                cfg.on_claim(r)

    async def _run_settle_job(self) -> None:
        cfg = self._auto_settle_config
        ctx = self._build_auto_settlement_context(now_ms())
        if not ctx.pending_settle:
            return
        if cfg.should_settle is not None:
            should = await _await_value(cfg.should_settle(ctx))
            if not should:
                return
        result = await self.settle()
        self._last_settle_time = now_ms()
        if cfg.on_settle is not None:
            cfg.on_settle(result)

    async def _run_refund_job(self) -> None:
        cfg = self._auto_settle_config
        if cfg.select_refund_channels is None:
            return
        ctx = self._build_auto_settlement_context(now_ms())
        channels = self._scheme.get_storage().list()
        targets = await _await_value(cfg.select_refund_channels(channels, ctx))
        for r in await self._refund_channels(targets):
            if cfg.on_refund is not None:
                cfg.on_refund(r)

    async def _select_claim_targets(self, opts: ClaimOptions | None) -> list[Channel]:
        channels = self._scheme.get_storage().list()
        if opts is None or opts.select_claim_channels is None:
            return channels
        ctx = self._build_auto_settlement_context(now_ms())
        return await _await_value(opts.select_claim_channels(channels, ctx))

    async def _claim_from_channels(
        self,
        channels: list[Channel],
        *,
        max_claims_per_batch: int,
        idle_secs: int | None,
    ) -> list[ClaimResult]:
        all_claims = get_claimable_vouchers_from_channels(channels, idle_secs)
        if not all_claims:
            return []
        results: list[ClaimResult] = []
        for i in range(0, len(all_claims), max_claims_per_batch):
            batch = all_claims[i : i + max_claims_per_batch]
            result = await self._submit_claim(batch)
            results.append(result)
            self._update_claimed_sessions(batch)
        if results:
            self._pending_settle = True
        return results

    async def _refund_channels(self, channels: list[Channel]) -> list[RefundResult]:
        now = now_ms()
        results: list[RefundResult] = []
        for c in channels:
            if has_live_pending_request(c, now):
                continue
            results.append(await self._refund_channel(c))
        return results

    async def _refund_channel(self, target: Channel) -> RefundResult:
        signer = self._scheme.get_receiver_authorizer_signer()
        claims = build_refund_claims(target)
        refund_amount = str(int(target.balance) - int(target.charged_cumulative_amount))
        nonce = str(target.refund_nonce or 0)

        refund_authorizer_signature = (
            sign_refund(signer, target.channel_id, refund_amount, nonce, self._network)
            if signer is not None
            else None
        )
        claim_authorizer_signature = (
            sign_claim_batch(signer, claims, self._network)
            if signer is not None and claims
            else None
        )

        requirements = build_payment_requirements(self._receiver, self._token, self._network)
        payment_payload = build_refund_payment_payload(
            target,
            claims,
            requirements,
            refund_amount,
            nonce,
            refund_authorizer_signature,
            claim_authorizer_signature,
        )

        response = await self._facilitator.settle(payment_payload, requirements)
        if not response.success:
            raise RuntimeError(format_facilitator_failure("Refund", response))

        def update(current: Channel | None) -> Channel | None:
            if current is None:
                return current
            if has_live_pending_request(current, now_ms()):
                return current
            return None

        self._scheme.get_storage().update_channel(target.channel_id, update)

        return RefundResult(channel=target.channel_id, transaction=response.transaction)

    async def _submit_claim(self, claims: list[VoucherClaim]) -> ClaimResult:
        signer = self._scheme.get_receiver_authorizer_signer()
        claim_authorizer_signature = (
            sign_claim_batch(signer, claims, self._network) if signer is not None else None
        )
        requirements = build_payment_requirements(self._receiver, self._token, self._network)
        payment_payload = build_claim_payment_payload(
            claims, requirements, claim_authorizer_signature
        )
        response = await self._facilitator.settle(payment_payload, requirements)
        if not response.success:
            raise RuntimeError(format_facilitator_failure("Claim", response))
        return ClaimResult(vouchers=len(claims), transaction=response.transaction)

    def _update_claimed_sessions(self, claims: list[VoucherClaim]) -> None:
        storage = self._scheme.get_storage()
        for claim in claims:
            channel_id = compute_channel_id(claim.channel, self._network)
            channel = storage.get(channel_id)
            if channel is None:
                continue
            claimed_amount = int(claim.total_claimed)
            if claimed_amount <= int(channel.total_claimed):
                continue

            def update(current: Channel | None, _claimed: int = claimed_amount) -> Channel | None:
                if current is None or _claimed <= int(current.total_claimed):
                    return current
                next_ch = current.copy()
                next_ch.total_claimed = str(_claimed)
                return next_ch

            storage.update_channel(channel_id, update)

    def _build_auto_settlement_context(self, now_ms_val: int) -> AutoSettlementContext:
        return AutoSettlementContext(
            now=now_ms_val,
            last_claim_time=self._last_claim_time,
            last_settle_time=self._last_settle_time,
            pending_settle=self._pending_settle,
        )


async def _await_value(value: Any | Awaitable[Any]) -> Any:
    if asyncio.iscoroutine(value) or asyncio.isfuture(value):
        return await value
    return value


__all__ = [
    "AutoSettlementConfig",
    "AutoSettlementContext",
    "BatchSettlementChannelManager",
    "ChannelManagerConfig",
    "ClaimChannelSelector",
    "ClaimOptions",
    "ClaimResult",
    "RefundChannelSelector",
    "RefundResult",
    "SettleResult",
]
